slyd/slyd/splash/proxy.py

```diff
@@ -20,31 +22,55 @@ def render_GET(self, request):
         url = request.args['url'][0]
         referer = request.args['referer'][0]
         tabid = int(request.args['tabid'][0])
-
         user = User.findById(tabid)
-        cb = functools.partial(self.end_response, request, url, tabid)
+
+        connection_status = { "finished": False }
+        cb = functools.partial(self.end_response, request, url, connection_status, tabid)
         if not user or not user.tab:
             d = deferToThread(requests.get, url, headers={'referer': referer})
             d.addCallback(cb)
+            d.addErrback(self._requestError, request)
+            request.notifyFinish().addErrback(self._requestDisconnect, deferred=d)
             return NOT_DONE_YET
 
         if request.auth_info['username'] != user.auth['username']:
             return self._error(request, 403, "You don't own that browser session")
 
+        request.notifyFinish().addErrback(self._requestDisconnect, None, connection_status)
         user.tab.http_client.get(url, cb, headers={'referer': referer})
         return NOT_DONE_YET
 
-    def end_response(self, request, original_url, tabid, reply):
+    def _requestError(self, err, request):
+        if not err.check(CancelledError):
+            request.setResponseCode(500)
+            request.write('Error fetching the content')
+            request.finish()
+
+    def _requestDisconnect(self, err, deferred=None, connection_status=None):
+        print "_requestDisconnect"
```

Comment:
For debug, forgot to remove it. Will update the PR.
On Jul 16, 2015 6:22 PM, "Ruairi Fahy" <notifications@github.com> wrote:

> In slyd/slyd/splash/proxy.py
> <https://github.com/scrapinghub/portia/pull/232#discussion_r34777580>:
>
> >          user.tab.http_client.get(url, cb, headers={'referer': referer})
> >          return NOT_DONE_YET
> >
> > -    def end_response(self, request, original_url, tabid, reply):
> > +    def _requestError(self, err, request):
> > +        if not err.check(CancelledError):
> > +            request.setResponseCode(500)
> > +            request.write('Error fetching the content')
> > +            request.finish()
> > +
> > +    def _requestDisconnect(self, err, deferred=None, connection_status=None):
> > +        print "_requestDisconnect"
>
> Is this for debug or to keep track of when this is called?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scrapinghub/portia/pull/232/files#r34777580>.
>